pkg/controller.v1/jax/jaxjob_controller.go

```diff
@@ -0,0 +1,485 @@
+			}
+		}
+
+		if failed > 0 && (specReplicas > succeeded+running) {
+			if spec.RestartPolicy != kubeflowv1.RestartPolicyNever {
+				msg := fmt.Sprintf("JAXJob %s is restarting because %d %s replica(s) failed.", jaxjob.Name, failed, rtype)
+				r.Recorder.Event(jaxjob, corev1.EventTypeWarning, commonutil.NewReason(kubeflowv1.JAXJobKind, commonutil.JobRestartingReason), msg)
+				commonutil.UpdateJobConditions(jobStatus, kubeflowv1.JobRestarting, corev1.ConditionTrue, commonutil.NewReason(kubeflowv1.JAXJobKind, commonutil.JobRestartingReason), msg)
+				trainingoperatorcommon.RestartedJobsCounterInc(jaxjob.Namespace, r.GetFrameworkName())
+			} else {
+				msg := fmt.Sprintf("JAXJob %s is failed because %d %s replica(s) failed.", jaxjob.Name, failed, rtype)
+				r.Recorder.Event(jaxjob, corev1.EventTypeNormal, commonutil.NewReason(kubeflowv1.JAXJobKind, commonutil.JobFailedReason), msg)
+				if jobStatus.CompletionTime == nil {
+					now := metav1.Now()
+					jobStatus.CompletionTime = &now
+				}
+				commonutil.UpdateJobConditions(jobStatus, kubeflowv1.JobFailed, corev1.ConditionTrue, commonutil.NewReason(kubeflowv1.JAXJobKind, commonutil.JobFailedReason), msg)
+				trainingoperatorcommon.FailedJobsCounterInc(jaxjob.Namespace, r.GetFrameworkName())
+			}
+		}
+	}
+	return nil
+}
+
+// UpdateJobStatusInApiServer updates the job status in to cluster.
+func (r *JAXJobReconciler) UpdateJobStatusInApiServer(job interface{}, jobStatus *kubeflowv1.JobStatus) error {
+	if jobStatus.ReplicaStatuses == nil {
+		jobStatus.ReplicaStatuses = map[kubeflowv1.ReplicaType]*kubeflowv1.ReplicaStatus{}
+	}
+
+	jaxjob, ok := job.(*kubeflowv1.JAXJob)
+	trainingoperatorcommon.ClearGeneratedFields(&jaxjob.ObjectMeta)
+	if !ok {
+		return fmt.Errorf("%+v is not a type of JAXJob", job)
+	}
+
+	// Job status passed in differs with status in job, update in basis of the passed in one.
+	if !equality.Semantic.DeepEqual(&jaxjob.Status, jobStatus) {
+		jaxjob = jaxjob.DeepCopy()
+		jaxjob.Status = *jobStatus.DeepCopy()
+	}
+
+	result := r.Status().Update(context.Background(), jaxjob)
+
+	if result != nil {
+		r.Log.WithValues("jaxjob", types.NamespacedName{
+			Namespace: jaxjob.GetNamespace(),
+			Name:      jaxjob.GetName(),
+		})
+		return result
+	}
+
+	return nil
+}
+
+// SetClusterSpec sets the cluster spec and init container for the pod
+func (r *JAXJobReconciler) SetClusterSpec(job interface{}, podTemplate *corev1.PodTemplateSpec, rtype, index string) error {
+	if err := setPodEnv(job, podTemplate, rtype, index); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *JAXJobReconciler) GetDefaultContainerName() string {
+	return kubeflowv1.JAXJobDefaultContainerName
+}
+
+func (r *JAXJobReconciler) GetDefaultContainerPortName() string {
+	return kubeflowv1.JAXJobDefaultPortName
+}
+
+func (r *JAXJobReconciler) IsMasterRole(replicas map[kubeflowv1.ReplicaType]*kubeflowv1.ReplicaSpec,
+	rtype kubeflowv1.ReplicaType, index int) bool {
+	return false
```

Comment:
```suggestion
	return index == 0
```

Index 0 indicates coordinator (master), right?